File: packages/core/src/serialize.ts
/**
 * Byte-stable serializer: Model → text. A given model always produces identical bytes.
 * Tables are compact (single-space cell padding, never alignment padding) and one row = one
 * line, so a single-cell edit maps to a minimal text splice.
 */

import { escapeCell } from './escape.js'
import type { Model, Sheet } from './model.js'
import { parse } from './parse.js'
import { serializeStyleBlock } from './style.js'

/**
 * Normalize arbitrary Defter text to its canonical byte-stable form. **Do this once before binding
 * text to a CRDT** — otherwise the first edit's serialize carries one-time normalization churn,
 * making the splice non-minimal and able to overlap a concurrent edit.
 */
export function normalize(text: string): string {
  return serialize(parse(text))
}

export function serialize(model: Model): string {
  const multi = model.sheets.length > 1
  const blocks = model.sheets.map((sheet) => serializeSheet(sheet, multi))
  return `${blocks.join('\n\n')}\n`
}

function serializeSheet(sheet: Sheet, forceHeading: boolean): string {
  const width = Math.max(sheet.width, 1)
  const parts: string[] = []
  if (sheet.headed || forceHeading) parts.push(`## Sheet: ${sheet.name}\n`)

  const row = (cells: string[]): string => {
    const padded = Array.from({ length: width }, (_, c) => escapeCell(cells[c] ?? ''))
    return `| ${padded.join(' | ')} |`
  }

  const header = sheet.grid[0] ?? []
  const delim = Array.from({ length: width }, (_, c) => alignMarker(sheet.colAlign[c] ?? null))
  const lines = [row(header), `| ${delim.join(' | ')} |`]
  for (let r = 1; r < sheet.grid.length; r++) lines.push(row(sheet.grid[r]!))
  parts.push(lines.join('\n'))

  const hasFreeze = sheet.freeze !== undefined && (sheet.freeze.rows > 0 || sheet.freeze.cols > 0)
  if (
    sheet.styles.length > 0 ||
    sheet.charts.length > 0 ||
    sheet.conditionals.length > 0 ||
    sheet.validations.length > 0 ||
    sheet.checkboxes.length > 0 ||
    sheet.dates.length > 0 ||
    sheet.filters.length > 0 ||
    sheet.names.length > 0 ||
    hasFreeze
  ) {
    const block = serializeStyleBlock(
      sheet.styles,
      sheet.charts,
      sheet.conditionals,
      sheet.validations,
      sheet.checkboxes,
      sheet.dates,
      sheet.filters,
      sheet.names,
      sheet.freeze,
    )
    parts.push(`\n\`\`\`defter-style\n${block}\n\`\`\``)
  }
  return parts.join('\n')
}

function alignMarker(align: 'left' | 'center' | 'right' | null): string {
  switch (align) {
    case 'left':
      return ':--'
    case 'center':
      return ':-:'
    case 'right':
      return '--:'
    default:
      return '---'
  }
}

File: go/serialize.go
package defterparse

import "strings"

// Byte-stable serializer: Model -> text. Mirrors packages/core/src/serialize.ts.
// A given model always produces identical bytes. Tables are compact (single-space
// cell padding, never alignment padding) and one row = one line.

// Normalize turns arbitrary Defter text into its canonical byte-stable form.
// Equivalent to Serialize(Parse(text)).
func Normalize(text string) string {
	return Serialize(Parse(text))
}

// Serialize renders a Model to canonical text.
func Serialize(model *Model) string {
	multi := len(model.Sheets) > 1
	blocks := make([]string, len(model.Sheets))
	for i, s := range model.Sheets {
		blocks[i] = serializeSheet(s, multi)
	}
	return strings.Join(blocks, "\n\n") + "\n"
}

func serializeSheet(sheet *Sheet, forceHeading bool) string {
	width := sheet.Width
	if width < 1 {
		width = 1
	}
	var parts []string
	if sheet.Headed || forceHeading {
		parts = append(parts, "## Sheet: "+sheet.Name+"\n")
	}

	row := func(cells []string) string {
		padded := make([]string, width)
		for c := 0; c < width; c++ {
			v := ""
			if c < len(cells) {
				v = cells[c]
			}
			padded[c] = escapeCell(v)
		}
		return "| " + strings.Join(padded, " | ") + " |"
	}

	var header []string
	if len(sheet.Grid) > 0 {
		header = sheet.Grid[0]
	}
	delim := make([]string, width)
	for c := 0; c < width; c++ {
		a := AlignNone
		if c < len(sheet.ColAlign) {
			a = sheet.ColAlign[c]
		}
		delim[c] = alignMarker(a)
	}
	lines := []string{row(header), "| " + strings.Join(delim, " | ") + " |"}
	for r := 1; r < len(sheet.Grid); r++ {
		lines = append(lines, row(sheet.Grid[r]))
	}
	parts = append(parts, strings.Join(lines, "\n"))

	if len(sheet.Styles) > 0 || len(sheet.Charts) > 0 || len(sheet.Conditionals) > 0 ||
		len(sheet.Validations) > 0 || len(sheet.Names) > 0 {
		block := serializeStyleBlock(sheet.Styles, sheet.Charts, sheet.Conditionals, sheet.Validations, sheet.Names)
		parts = append(parts, "\n```defter-style\n"+block+"\n```")
	}
	return strings.Join(parts, "\n")
}

func alignMarker(a Align) string {
	switch a {
	case AlignLeft:
		return ":--"
	case AlignCenter:
		return ":-:"
	case AlignRight:
		return "--:"
	default:
		return "---"
	}
}
